### DISCUSSION TITLE
Source helper scripts from run_ scripts possible ?

### DISCUSSION THREAD

**Discussion by oschrenk** [twpayne/chezmoi#3764] [General]
I have a bunch of configuration scripts inside `.chezmoiscripts`

Let's say

```
run_onchange_03_configure_apps_chrome.sh
run_onchange_03_configure_apps_flux.sh
run_onchange_03_configure_apps_hammerspoon.sh
run_onchange_03_configure_apps_homerow.sh
...
```

I like to keep them separate for organizational purposes. 
But some of the code inside is usually the same, so I extracted it into

`configure_apps__helper.sh`

and source it via

```
BASEDIR=$(dirname $0)
source $BASEDIR/configure_apps__helper.sh
```

It seems though that chezmoi copies the template before executing it to a temporary file, which of course then doesn't contain the helper script.

Apart from hardcoding the path to the helper script is there another way sharing code between the scripts? Ideas?

**Comment by bradenhilton** [comment]
Where is `configure_apps__helper.sh` located?

You can get the source directory with `chezmoi source-path`, or you can turn the scripts into templates and use `{{ .chezmoi.sourceDir }}`.

**Reply by oschrenk** [reply]
It's next to the other scripts in `$(chezmoi source-path)/.chezmoiscripts`

Using` chezmoi source-path` sounds like the easiest way forward. Don't want to convert all scripts into templates just for that.

Thank you!

**Comment by RedX2501** [comment]
I'm not able to make this work.

Not sure if something changed since the post above, but now keeping any non script file (how exactly does it know what a script file is?) in the `.chezmoiscripts` folder results in the error: `chezmoi: /home/⟨user⟩/dotfiles/home/.chezmoiscripts/configure_app_lib_xdg.sh: not a script`.

I tried making it an ignored file (reading the source code tells me those start with a `.`) leads to it not being copied.

What is the correct approach for version `2.69.4` without resorting to templates?

**Reply by halostatue** [reply]
Templates.

**Reply by RedX2501** [reply]
Templates for shell scripts are quite annoying as shellcheck is no longer able to provide static code checks for them. Even for small scripts that is useful, especially when writing for `sh` instead of a more familiar format.

My workaround is to use an environment variable.

**Reply by RedX2501** [reply]
After playing around a bit more I think this is even better. Yes it involves another call to chezmoi but it seems to work fine:

```
cmSourceDir="$(chezmoi execute-template '{{ .chezmoi.sourceDir }}')"
. "$cmSourceDir/.chezmoiscripts/.lib_xdg.sh"
```

**Reply by bradenhilton** [reply]
You asked for the correct approach without resorting to templates, but this isn't a recommended approach, so the question isn't really answerable. The correct/recommended approach is to use templates, *especially* for chezmoi-specific values. What you're currently doing is essentially rolling your own version of the built in templating with extra steps.

I would be shocked if your editor and/or shellcheck couldn't be configured in some way to treat `*.sh.tmpl` files `*.sh` files.

**Reply by RedX2501** [reply]
I understand that the recommended way is to use templates. I think that others, like I, would like to have "pure" shellscripts, and will be looking for a solution.

So the second comment was more in the spirit of documenting the workaround I found.

Yes, VS Code with shellcheck can be configured to treat .sh.tmpl as shellscript files.

The downside is that you get shellcheck warnings as the syntax is not really compatible:

<img width="756" height="543" alt="image" src="https://github.com/user-attachments/assets/10b01bd4-6624-42ff-a61d-90c0d78ec57f" />

I do realize that for the template in the string there is no warning. So in some cases it might work out well or that could be used to work around the incompatible syntax.

Why I had to use the more convoluted `execute-template` version is because `chezmoi source-path` is always empty in the scripts. It does not matter whether in shell or templates.

The source file is the same, once as template and once as shell script.
```sh
echo "cz sd $(chezmoi source-path)"
echo "cz et $(chezmoi execute-template '{{ .chezmoi.sourceDir }}')"
echo "template {{ .chezmoi.sourceDir }}"
```

Output as script:
```
~$ chezmoi --source dotfiles-orig/ init
~$ chezmoi --source dotfiles-orig/ apply
cz sd 
cz et /home/gui/dotfiles-orig/home
template {{ .chezmoi.sourceDir }}
```

(Of course in the shell file the template does not get replaced. I just kept it for completeness.)

Output as template:
```
~$ chezmoi --source dotfiles-orig/ init
~$ chezmoi --source dotfiles-orig/ apply
cz sd 
cz et /home/gui/dotfiles-orig/home
template /home/gui/dotfiles-orig/home
```

Running it on it's own does work:
```sh
~$ chezmoi source-path 
/home/gui/dotfiles-orig/home
```

**Reply by halostatue** [reply]
You cannot safely use execute-template while running apply.

Look at my dotfiles to see how to do this with shfmt and shellcheck working just fine.

The answer is templates.

**Reply by bradenhilton** [reply]
You can [configure the delimiters per-template using a directive](https://www.chezmoi.io/reference/templates/directives/#delimiters) to fix the syntax errors, I typically just prepend a language-specific comment sequence to the left delimiter so it appears as a comment in the template.

`/path/to/script.sh.tmpl`:
```text
# chezmoi:template:left-delimiter="#{{"

...

#{{ if eq .chezmoi.sourceDir "test" }}
...
#{{ end }}
```

**Reply by bradenhilton** [reply]
Also, if you end up using templates, you *can* still source the files in the script, but you can also just opt to include the contents of the modular snippets/fragments directly with [`include`](https://www.chezmoi.io/reference/templates/functions/include/) (or [`includeTemplate`](https://www.chezmoi.io/reference/templates/functions/includeTemplate/), if the fragment you are including is itself also a template):

```text
#!/bin/sh

# chezmoi:template:left-delimiter="#{{"

#{{ include ".chezmoiscripts/.lib_xdg.sh" }}

...
```

(I'm not on a UNIX-like machine right now, so check the above for correctness)

This approach is more robust and can signal potential issues early on during diffs (if the file no longer exists, the script contents will be different), which you don't get when sourcing files, they just blow up when you run them.

For what it's worth, when I do this, I put all my snippets/fragments in `.chezmoitemplates` (such as `.chezmoitemplates/scripts/sh/lib_xdg.sh`, but you can name them as you wish) and use `includeTemplate` instead. You also don't have to worry about ignoring the fragment files if you do this.
